### PR TITLE
Add support for actuators requeuing machines after duration.

### PR DESCRIPTION
**Pull request by dgoodwin**
For some actuators we may get our expected cloud instance created, but
it will take a little time before we can populate all the status we want
to from it. (i.e. running state, public IP, etc) This results in the machine 
status showing pending until the next resync in ~10 minutes on AWS 
for example.

I investigated simply returning an error, but this results in the max retries of 10 being hit very
quickly with the current generated controller code and exponential
backoff settings. Because this code appears to be auto-generated, 
I'm not sure if we can or want to start editing it to slow down the retries.

This change adds a specific error type actuators can return to request
that the machine be re-synced in some duration.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by sidharthsurana** [comment]
@dgoodwin Thanks for putting this together. I have also run into the same issue for the vsphere provider, where just 10 tries at times are not sufficient. The approach in your PR seems to be a very reasonable.

File: pkg/controller/machine/controller.go [hunk 1]
```diff
@@ -34,6 +38,14 @@ import (
 
 const NodeNameEnvVar = "NODE_NAME"
 
+type RequeueAfterError struct {
+	RequeueAfter time.Duration
+}
+
+func (e *RequeueAfterError) Error() string {
+	return fmt.Sprintf("Requeue machine in: %s", e.RequeueAfter)
```
Comment (roberthbailey) [inline pkg/controller/machine/controller.go]: nit: should start with a lower case letter.

File: pkg/controller/machine/controller.go [hunk 2]
```diff
@@ -128,14 +143,38 @@ func (c *MachineControllerImpl) Reconcile(machine *clusterv1.Machine) error {
 	}
 	if exist {
 		glog.Infof("Reconciling machine object %v triggers idempotent update.", name)
-		return c.update(m)
+		err = c.update(m)
+		if err != nil {
+			requeueErr, ok := err.(*RequeueAfterError)
```
Comment (roberthbailey) [inline pkg/controller/machine/controller.go]: nit: combine with the next line:

`if requeueErr, ok := err.(*RequeueAfterError); ok { ... }`

File: pkg/controller/machine/controller.go [hunk 3]
```diff
@@ -128,14 +143,38 @@ func (c *MachineControllerImpl) Reconcile(machine *clusterv1.Machine) error {
...
+			}
+		}
+		return err
 	}
 	// Machine resource created. Machine does not yet exist.
 	glog.Infof("Reconciling machine object %v triggers idempotent create.", m.ObjectMeta.Name)
 	if err := c.create(m); err != nil {
-		glog.Warningf("unable to create machine %v: %v", name, err)
+		glog.Warningf("Unable to create machine %v: %v", name, err)
+		requeueErr, ok := err.(*RequeueAfterError)
```
Comment (roberthbailey) [inline pkg/controller/machine/controller.go]: same here:

`if requeueErr, ok := err.(*RequeueAfterError); ok { ... }`

File: pkg/controller/machine/controller.go [hunk 4]
```diff
@@ -128,14 +143,38 @@ func (c *MachineControllerImpl) Reconcile(machine *clusterv1.Machine) error {
...
+		if ok {
+			glog.Infof("Actuator returned requeue-after error: %v", requeueErr)
+			return c.enqueueAfter(machine, requeueErr.RequeueAfter)
+		} else {
```
Comment (roberthbailey) [inline pkg/controller/machine/controller.go]: You don't need an else here since the if block returns.

File: pkg/controller/machine/controller.go [hunk 5]
```diff
@@ -128,14 +143,38 @@ func (c *MachineControllerImpl) Reconcile(machine *clusterv1.Machine) error {
...
+	return nil
+}
+
+func (c *MachineControllerImpl) enqueueAfter(machine *clusterv1.Machine, after time.Duration) error {
+	key, err := cache.MetaNamespaceKeyFunc(machine)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", machine, after)
 		return err
 	}
+	c.queue.Queue.AddAfter(key, after)
```
Comment (roberthbailey) [inline pkg/controller/machine/controller.go]: Does the queue have infinite size or can this call return an error?

Comment (dgoodwin) [reply pkg/controller/machine/controller.go]: Looks like it must be infinite, no return values from the function.

**Comment by dgoodwin** [comment]
Comments addressed thanks for giving it a look @roberthbailey .

**Pull request merged**